### PR TITLE
Support reading and writing DICOM whole slide images

### PR DESCRIPTION
**Pull request by melissalinkert**
Significantly reworks the DICOM reader and adds a new writer so that WSI data is supported.

Relevant test data is in ```inbox/dicom-wsi```.  There are a ton of datasets there, but in general the newer sets (```LUNG...```, ```pixelmed-converted```, ```WG26/3DHISTECH```, ```WG26/Leica_GT450```) will be better for testing by hand.  Most are copied from ```ftp://dicom.nema.org/MEDICAL/Dicom/DataSets/WG26```, so publicly available.

Whole slide datasets will have one or more files per resolution, with optional additional files that contain the label, overview, etc.  Multiple pyramids per dataset are possible if an extended depth of field image was created from a Z stack.  Both brightfield and fluorescence slides are supported.

Converting non-DICOM slides to DICOM is supported by the new ```DicomWriter```, using typical ```bfconvert``` options.  The expected output is one file per resolution or extra (label/macro/etc.) image.  Some changes were needed in ```ImageConverter``` to make the tile writing logic less TIFF-specific.

There is also a new ```bfconvert``` option to turn off the sequential-writing assumption (https://github.com/ome/bioformats/commit/3208d92e73bb5ba78bed31cd15c41dacf076b21d).  This doesn't change the actual  tile writing order, just the assumption that the writer uses.  For DICOM, this is helpful for testing the two tile position storage modes: ```TILED_FULL``` and ```TILED_SPARSE```.  The former implicitly defines the positions; all tiles are stored, and left-to-right, top-to-bottom.  The latter explicitly records positions, which results in a larger file (more metadata), but allows missing tiles and/or non-sequential storage.

There is obviously a lot to go through here, so happy to have a discussion if that's helpful.

This will definitely impact memo files so should not be considered for a patch release.  Most test issues will be addressed in a forthcoming config PR, and are primarily updates to image names/descriptions and a few pixel hash changes due to differences in file sorting (by metadata now instead of by file name/pattern).  There are a couple of specific files which will still cause failures, and a proposed solution will be noted on the config PR.

I have it on my to-do list to go through outstanding DICOM issues to verify whether or not this helps.  In particular:
  - #3509
  - #3619
  - #3620
  - #3622
  - https://trello.com/c/SeNEOVDN/360-dicom-add-support-for-parent-child-relationship-in-metadata
  - https://trello.com/c/f4r7sbke/240-dicom-corrupted-images

### CODE REVIEW

File: components/formats-bsd/src/loci/formats/in/DicomTag.java [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+        int len = DataTools.bytesToInt(b, 0, 4, in.isLittleEndian());
+        if (len == 0xffffffff) {
+          return len;
+        }
+        else {
+          vr = IMPLICIT;
+          return 8;
+        }
+      default:
+        throw new IllegalArgumentException(vr.toString());
```
Comment (sbesson) [inline components/formats-bsd/src/loci/formats/in/DicomTag.java:452]: As a result of the DICOM reader priority increase in `readers.txt`, this exception is now being thrown across several file formats during the `isThisType` check.

See e.g. [bd-pathway](https://merge-ci.openmicroscopy.org/jenkins/job/BIOFORMATS-test-folder/153391/console), [deltavision](https://merge-ci.openmicroscopy.org/jenkins/job/BIOFORMATS-test-folder/153394/console) for examples of affected filesets.

**Comment by fedorov** [comment]
@melissalinkert would it be possible to acknowledge that development of this functionality was supported by NCI Imaging Data Commons, and include the wording for the funding source?

**Comment by melissalinkert** [comment]
@fedorov : we can certainly add a funding acknowledgement. Typically we do this in the release notes via a pull request to https://github.com/ome/bio-formats-documentation.  An example is the ```New file formats``` section of the 6.6.0 notes:

https://docs.openmicroscopy.org/bio-formats/6.7.0/about/whats-new.html#december

If you have specific acknowledgement text you prefer, or a line in the release notes is insufficient, just let us know.

**Comment by fedorov** [comment]
Melissa, thank you for the pointer, this is very helpful. Let me check with your project leads at NCI/Leidos what is their preference, and I will get back to you towards the end of the week.

File: components/formats-bsd/src/loci/formats/in/DicomReader.java [hunk 2]
```diff
@@ -65,78 +66,30 @@
 import ome.units.quantity.Length;
 import ome.units.UNITS;
 
+import static loci.formats.in.DicomAttribute.*;
+import static loci.formats.in.DicomVR.*;
```
Comment (sbesson) [inline components/formats-bsd/src/loci/formats/in/DicomReader.java:70]: As mentioned during the weekly Formats discussion, immediate question is whether this API handling the specifics of the DICOM specification should be managed in a dedicated package similarly to `loci.formats.tiff`.

Comment (melissalinkert) [reply components/formats-bsd/src/loci/formats/in/DicomReader.java]: 625f8cc reorganizes into a new ```loci.formats.dicom``` package.

File: components/formats-bsd/src/loci/formats/in/DicomTag.java (hunk 1 above)
Comment (melissalinkert) [reply components/formats-bsd/src/loci/formats/in/DicomTag.java]: I'd expect f75b8dc, 84e1a80, and fc13799 to resolve unexpected test failures, but will check the results tomorrow now that builds are running again.

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [BIOFORMATS-push#1041](https://merge-ci.openmicroscopy.org/jenkins/job/BIOFORMATS-push/1041/). See the [console output](https://merge-ci.openmicroscopy.org/jenkins/job/BIOFORMATS-push/1041/consoleText) for more details.
Possible conflicts:
  - PR #3709 jmuhlich 'Use multiple-of-16 tile dimensions when writing TIFFs'
    - components/bio-formats-tools/src/loci/formats/tools/ImageConverter.java

--conflicts

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [BIOFORMATS-push#1042](https://merge-ci.openmicroscopy.org/jenkins/job/BIOFORMATS-push/1042/). See the [console output](https://merge-ci.openmicroscopy.org/jenkins/job/BIOFORMATS-push/1042/consoleText) for more details.
Possible conflicts:
  - PR #3709 jmuhlich 'Use multiple-of-16 tile dimensions when writing TIFFs'
    - components/bio-formats-tools/src/loci/formats/tools/ImageConverter.java

--conflicts

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [BIOFORMATS-push#1049](https://merge-ci.openmicroscopy.org/jenkins/job/BIOFORMATS-push/1049/). See the [console output](https://merge-ci.openmicroscopy.org/jenkins/job/BIOFORMATS-push/1049/consoleText) for more details.
Possible conflicts:
  - PR #3709 jmuhlich 'Use multiple-of-16 tile dimensions when writing TIFFs'
    - components/bio-formats-tools/src/loci/formats/tools/ImageConverter.java

--conflicts

**Comment by fedorov** [comment]
Melissa, I am not sure how documentation updates are handled - would that be done in a separate repo after this PR is merged, to document improvements to DICOM support?

**Comment by fedorov** [comment]
Melissa, regarding the acknowledgment, I checked with our federal leads, and they would like to have the following acknowledgment, if possible:

"This project has been funded in whole or in part with Federal funds from the National Cancer Institute,
National Institutes of Health, under Task Order No. HHSN26110071 under Contract No. HHSN2612015000031."

cc: @ulrikew

**Comment by dgault** [comment]
Builds and tests have been green with this PR included and the merge-ci repo tests are now green also with the associated config. Overall the PR looks to be in good shape and suitable for inclusion with the 6.8.0 release.

While testing the writer I noticed a few times that converted files would produce the below exception. This appears to have been due to files being grouped with others I was testing, when the files were moved to a separate directory they opened without exception. This is likely a similar issue that required some of the config files to be moved.
```
Exception in thread "main" java.lang.ArithmeticException: / by zero
	at loci.formats.MetadataTools.populatePixelsOnly(MetadataTools.java:315)
	at loci.formats.MetadataTools.populateMetadata(MetadataTools.java:262)
	at loci.formats.MetadataTools.populatePixels(MetadataTools.java:151)
	at loci.formats.MetadataTools.populatePixels(MetadataTools.java:116)
	at loci.formats.in.DicomReader.initFile(DicomReader.java:905)
```

I manually tested the reader across a subset of the new sample files:

`CPTAC-LSCC_C3L-00965-26`
Dataset opened and displayed without exception, containing 1 series with 8 resolutions.  

`CPTAC-LSCC_C3L-00965-26_PixelMedConvertedSVSSuppliedPyramid`
Dataset opened and displayed without exception, containing 3 series, the first with 3 resolutions, then a thumbnail and overview image.  However when attempting to open the file `CPTAC-LSCC_C3L-00965-26_DICOMDIR` an exception was thrown as below. Based on the config PR this file was set with test=false so I assume it is not expected to be able to open this file.
```
java.lang.ArithmeticException: / by zero
	at loci.formats.in.DicomReader.initFile(DicomReader.java:719)
```

`LUNG-1-LN^ [LUNG-1-LN]`
Each series could be opened and displayed without exceptions. Each had 5 resolutions.

`WG26/3DHISTECH/FL_20191127/DICOM_TILED_FULL/056_ATM`
The dataset was able to opened and displayed without exception. It contained 4 series. The first series has 10 resolutions then an overview image, another image with 10 resolutions and the final series is labelled REGIONLOCALIZER

`WG26/3DHISTECH/FL_20191127/DICOM_TILED_SPARSE/056_ATM`
The dataset was able to opened and displayed without exception matching that of the TILED_FULL dataset 

`WG26/3DHISTECH/FL_20191127/DICOM_TILED_FULL/FluorCell_4`
The dataset was able to opened and displayed without exception. It contained 3 series. The first series has 10 resolutions then an image REGIONLOCALIZER followed by a label image.

`WG26/3DHISTECH/FL_20191127/DICOM_TILED_SPARSE/FluorCell_4`
The dataset was able to opened and displayed without exception matching that of the TILED_FULL dataset

`WG26/3DHISTECH/FL_20191127/DICOM_TILED_FULL/Liver_v1`
The dataset was able to opened and displayed without exception. It contained 4 series. The first series has 10 resolutions then an overview image, another image with 10 resolutions and the final series is labelled REGIONLOCALIZER

`WG26/3DHISTECH/FL_20191127/DICOM_TILED_SPARSE/Liver_v1`
The dataset was able to opened and displayed without exception matching that of the TILED_FULL dataset 

`WG26/3DHISTECH/FL_20191127/DICOM_TILED_FULL/17361-13-FISH`
The dataset was able to opened and displayed without exception. It contained 4 series. The first series has 10 resolutions then an overview image, another image with 10 resolutions and the final series is labelled REGIONLOCALIZER

`WG26/3DHISTECH/FL_20191127/DICOM_TILED_SPARSE/17361-13-FISH`
The dataset was able to opened and displayed without exception matching that of the TILED_FULL dataset 



For the writer I tested the bfconvert workflow against a range of our public data.

For the below datasets with multiple timepoints a format exception was thrown. This is as expected:

- https://downloads.openmicroscopy.org/images/Leica-LIF/michael/
- https://downloads.openmicroscopy.org/images/CellH5/idr0013/

```
Exception in thread "main" loci.formats.FormatException: Multiple timepoints not supported
	at loci.formats.out.DicomWriter.setId(DicomWriter.java:335)

```
For other datasets the conversion was run both with and without the new option. Each of the resulting converted files opened and displayed as expected without any exceptions:

- https://downloads.openmicroscopy.org/images/OME-TIFF/2016-06/tubhiswt-2D/
- https://downloads.openmicroscopy.org/images/OME-TIFF/2016-06/BBBC/
- https://downloads.openmicroscopy.org/images/ND2/maxime/
- https://downloads.openmicroscopy.org/images/Olympus-OIR/etienne/
- https://downloads.openmicroscopy.org/images/Leica-SCN/openslide/Leica-1/


When running the converter using the -no-sequential across a range of the public sample data the resulting files matched those converted without the flag. However when running the converter with the flag and trying to generate resolutions I would run into the below exception (the index is 1 greater than the num planes):

`bfconvert -no-sequential -pyramid-resolutions 2 -pyramid-scale 2`
```
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 511 out of bounds for length 511
	at loci.formats.out.PyramidOMETiffWriter.close(PyramidOMETiffWriter.java:128)
	at loci.formats.ImageWriter.close(ImageWriter.java:469)
	at loci.formats.tools.ImageConverter.testConvert(ImageConverter.java:758)
However this error does not appear to be related to the changes in the ImageConverter code or the new option, it simply makes it easier to occur. You can achieve the same error using some of our example code and setting sequential writing to false. This is a bug that should probably be logged as a separate issue.
```

**Comment by dgault** [comment]
@fedorov, the addition to the version history (which will include the funding acknowledgement) mentioned by Melissa will be opened as a separate PR against https://github.com/ome/bio-formats-documentation. This usually happens once all of the PR's for a given release have been merged and the contents of that PR is then also used as the basis for the public announcements on our website and image.sc etc . 

If there are any other areas of documentation you feel need updated please do let me know.

**Comment by melissalinkert** [comment]
Thanks for the thorough review, @dgault.

> While testing the writer I noticed a few times that converted files would produce the below exception. This appears to have been due to files being grouped with others I was testing, when the files were moved to a separate directory they opened without exception. This is likely a similar issue that required some of the config files to be moved.

I definitely wouldn't recommend trying to put multiple converted DICOM pyramids in the same directory, since the file grouping is logic is mostly best guess.  Would it help to have the writer issue a warning if it will be writing to a non-empty directory?

> CPTAC-LSCC_C3L-00965-26_PixelMedConvertedSVSSuppliedPyramid
Dataset opened and displayed without exception, containing 3 series, the first with 3 resolutions, then a thumbnail and overview image. However when attempting to open the file CPTAC-LSCC_C3L-00965-26_DICOMDIR an exception was thrown as below. Based on the config PR this file was set with test=false so I assume it is not expected to be able to open this file.

Correct, I would not expect using the ```*DICOMDIR``` file for this particular dataset to work.

**Comment by dgault** [comment]
Thanks Melissa, getting this merged for now for 6.8.0

**Comment by dclunie** [comment]
@melissalinkert does this include support for DICOM z-stack images? E.g., see discussion at [news:comp.protocols.dicom](https://groups.google.com/g/comp.protocols.dicom/c/j0YAu_Jtnq0).

**Comment by melissalinkert** [comment]
@dclunie : yes, Z stacks are expected to be supported here. Several of the WG26 datasets referenced above include Z stacks, and those are included in nightly regression tests. If you (or anyone else following) have Z stacks that aren't being correctly read or written by Bio-Formats, we'll definitely want to know so that we can investigate the issue.

**Pull request merged**